syntax = "proto2";
package tzrec.protos;

import "tzrec/protos/module.proto";
import "tzrec/protos/tower.proto";

message SimpleMultiTask {
    repeated TaskTower task_towers = 1;
}

message MMoE {
    // mmoe expert module definition
    required MLP expert_mlp = 1;
    // mmoe gate module definition
    optional MLP gate_mlp = 2;
    // number of mmoe experts
    required uint32 num_expert = 3 [default=3];
    // task tower
    repeated TaskTower task_towers = 4;
}

message DBMTL {
    // shared bottom mlp layer
    optional MLP bottom_mlp = 1;

    // mmoe expert mlp layer definition
    optional MLP expert_mlp = 2;
    // mmoe gate module definition
    optional MLP gate_mlp = 3;
    // number of mmoe experts
    optional uint32 num_expert = 4 [default=3];

    // bayes task tower
    repeated BayesTaskTower task_towers = 5;
}

message DC2VR {
    // shared bottom mlp layer
    optional MLP bottom_mlp = 1;
    // mmoe expert mlp layer definition
    optional MLP expert_mlp = 2;
    // mmoe gate module definition
    optional MLP gate_mlp = 3;
    // number of mmoe experts
    optional uint32 num_expert = 4 [default=3];
    // task tower
    repeated InterventionTaskTower task_towers = 5;
}



message PLE {
    // extraction network
    repeated ExtractionNetwork extraction_networks = 1;
    // task tower
    repeated TaskTower task_towers = 2;
}


message DlrmHSTU {
    required string uih_id_feature_name = 11;

    required string uih_action_time_feature_name = 12;

    required string uih_action_weight_feature_name = 13;

    required string uih_watchtime_feature_name = 17;

    required string candidates_id_feature_name = 14;

    required string candidates_query_time_feature_name = 18;

    required string candidates_action_weight_feature_name = 19;

    required string candidates_watchtime_feature_name = 20;

    map<string, uint32> contextual_feature_to_max_length = 15;

    map<string, uint32> contextual_feature_to_min_uih_length = 16;

    map<string, uint32> action_weights = 21;

    optional uint32 hstu_num_heads = 1 [default = 4];

    optional uint32 hstu_attn_linear_dim = 2 [default = 128];

    optional uint32 hstu_attn_qk_dim = 3 [default = 128];

    optional uint32 hstu_attn_num_layers = 4 [default = 3];

    optional float hstu_input_dropout_ratio = 5 [default = 0.2];

    optional float hstu_linear_dropout_rate = 6 [default = 0.2];

    optional uint32 hstu_transducer_embedding_dim = 7 [default = 512];

    optional bool hstu_group_norm = 8 [default = false];
    //
    optional bool enable_postprocessor = 9 [default = true];

    optional bool use_layer_norm_postprocessor = 10 [default = false];

    // multi task tower config
    required FusionTaskTower fusion_task_tower = 22;

}
